### PR TITLE
pctl: Rework how pctl works to be more accurate

### LINKED ISSUE #6111: Unimplemented function 'IsFreeCommunicationAvailable'
Used by Monster Hunter Rise:

Debug <Critical> core\hle\service\service.cpp:ReportUnimplementedFunction:148: Assertion Failed! Unknown / unimplemented function 'IsFreeCommunicationAvailable': port='IParentalControlService' cmd_buf={[0]=0x6, [1]=0xC, [2]=0x0, [3]=0x0, [4]=0x100001, [5]=0x2, [6]=0x0, [7]=0x0, [8]=0x49434653}

### PR DESCRIPTION
**Pull request by ogniK5377**
Introduces the usage of compatibilities to allow the module to be closer to how it works on hardware.

Closes #6111

### CODE REVIEW

File: src/core/hle/service/pctl/module.cpp [hunk 1]
```diff
@@ -119,92 +133,273 @@ class IParentalControlService final : public ServiceFramework<IParentalControlSe
     }
 
 private:
+    bool CheckFreeCommunicationPermissionImpl() {
```
Comment (lioncash) [inline src/core/hle/service/pctl/module.cpp:136]:
```suggestion
    bool CheckFreeCommunicationPermissionImpl() const {
```

Also should this function be returning true on all paths, or is that resolved along with the completion of the TODO comment? If so, it's probably best to explicitly indicate in a comment that returning by true on all paths is currently intended and not a bug.

File: src/core/hle/service/pctl/module.cpp [hunk 2]
```diff
@@ -119,92 +133,273 @@ class IParentalControlService final : public ServiceFramework<IParentalControlSe
...
+            return true;
+        }
+        if (!settings.is_free_communication_default_on) {
+            return true;
+        }
+        // TODO(ogniK): Check for blacklisted/exempted applications
+        return true;
+    }
+
+    bool ConfirmStereoVisionPermissionImpl() {
```
Comment (lioncash) [inline src/core/hle/service/pctl/module.cpp:153]:
```suggestion
    bool ConfirmStereoVisionPermissionImpl() const {
```

File: src/core/hle/service/pctl/module.cpp [hunk 3]
```diff
@@ -119,92 +133,273 @@ class IParentalControlService final : public ServiceFramework<IParentalControlSe
...
+        LOG_DEBUG(Service_PCTL, "called");
+        IPC::ResponseBuilder rb{ctx, 2};
+
+        if (False(capability & (Capability::Application | Capability::System))) {
+            LOG_ERROR(Service_PCTL, "Invalid capability! capability={:X}",
+                      static_cast<s32>(capability));
+            return;
+        }
+
+        // TODO(ogniK): Recovery
```
Comment (lioncash) [inline src/core/hle/service/pctl/module.cpp:187]: Maybe expand this comment a little so its more explanatory what is meant by recovery.

File: src/core/hle/service/pctl/module.h [hunk 4]
```diff
@@ -12,19 +13,33 @@ class System;
 
 namespace Service::PCTL {
 
+enum class Capability : s32 {
```
Comment (lioncash) [inline src/core/hle/service/pctl/module.h:16]: Given this is operating on the enum as a set of bitflags, it's likely more appropriate here to make the backing type a `u32`

File: src/core/hle/service/pctl/module.h [hunk 5]
```diff
@@ -12,19 +13,33 @@ class System;
...
+    None = 0x0,
```
Comment (lioncash) [inline src/core/hle/service/pctl/module.h:17]:
```suggestion
    None = 0,
```

for consistency, given no other entries make use of hex literals.

File: src/core/hle/service/pctl/module.cpp [hunk 6]
```diff
@@ -119,92 +133,273 @@ class IParentalControlService final : public ServiceFramework<IParentalControlSe
...
+    }
+
     void Initialize(Kernel::HLERequestContext& ctx) {
-        LOG_WARNING(Service_PCTL, "(STUBBED) called");
+        LOG_DEBUG(Service_PCTL, "called");
+        IPC::ResponseBuilder rb{ctx, 2};
+
+        if (False(capability & (Capability::Application | Capability::System))) {
+            LOG_ERROR(Service_PCTL, "Invalid capability! capability={:X}",
+                      static_cast<s32>(capability));
```
Comment (lioncash) [inline src/core/hle/service/pctl/module.cpp:183]:
```suggestion
                      capability);
```

fmt handles passed in enum class values, so casts are no longer necessary.

File: src/core/hle/service/pctl/module.h [hunk 7]
```diff
@@ -12,19 +13,33 @@ class System;
...
+    Application = 1 << 0,
+    SnsPost = 1 << 1,
+    Recovery = 1 << 6,
+    Status = 1 << 8,
+    SteroVision = 1 << 9,
```
Comment (lioncash) [inline src/core/hle/service/pctl/module.h:22]:
```suggestion
    StereoVision = 1 << 9,
```

File: src/core/hle/service/pctl/module.cpp [hunk 8]
```diff
@@ -3,16 +3,30 @@
...
 #include "core/hle/service/pctl/module.h"
 #include "core/hle/service/pctl/pctl.h"
 
 namespace Service::PCTL {
 
+namespace Error {
+
+constexpr ResultCode ResultNoFreeCommunication{ErrorModule::PCTL, 101};
+constexpr ResultCode ResultStereoVisionRestricted{ErrorModule::PCTL, 104};
+constexpr ResultCode ResultNoCapatability{ErrorModule::PCTL, 131};
```
Comment (lioncash) [inline src/core/hle/service/pctl/module.cpp:20]:
```suggestion
constexpr ResultCode ResultNoCapability{ErrorModule::PCTL, 131};
```

File: src/core/hle/service/pctl/module.cpp [hunk 9]
```diff
@@ -119,92 +133,273 @@ class IParentalControlService final : public ServiceFramework<IParentalControlSe
...
+        rb.Push(pin_code[0] != '\0');
+    }
+
+    void ConfirmStereoVisionRestrictionConfigurable(Kernel::HLERequestContext& ctx) {
+        LOG_DEBUG(Service_PCTL, "called");
+
+        IPC::ResponseBuilder rb{ctx, 2};
+
+        if (False(capability & Capability::SteroVision)) {
+            LOG_ERROR(Service_PCTL, "Application does not have SteroVision capability!");
```
Comment (lioncash) [inline src/core/hle/service/pctl/module.cpp:268]:
```suggestion
            LOG_ERROR(Service_PCTL, "Application does not have StereoVision capability!");
```

File: src/core/hle/service/pctl/module.cpp [hunk 10]
```diff
@@ -119,92 +133,273 @@ class IParentalControlService final : public ServiceFramework<IParentalControlSe
...
         IPC::RequestParser rp{ctx};
         const auto can_use = rp.Pop<bool>();
-        LOG_WARNING(Service_PCTL, "(STUBBED) called, can_use={}", can_use);
-
-        can_use_stereo_vision = can_use;
+        LOG_DEBUG(Service_PCTL, "called, can_use={}", can_use);
 
         IPC::ResponseBuilder rb{ctx, 2};
+        if (False(capability & Capability::SteroVision)) {
+            LOG_ERROR(Service_PCTL, "Application does not have SteroVision capability!");
```
Comment (lioncash) [inline src/core/hle/service/pctl/module.cpp:301]:
```suggestion
            LOG_ERROR(Service_PCTL, "Application does not have StereoVision capability!");
```

File: src/core/hle/service/pctl/module.cpp [hunk 11]
```diff
@@ -119,92 +133,273 @@ class IParentalControlService final : public ServiceFramework<IParentalControlSe
...
         rb.Push(RESULT_SUCCESS);
     }
 
     void GetStereoVisionRestriction(Kernel::HLERequestContext& ctx) {
-        LOG_WARNING(Service_PCTL, "(STUBBED) called");
+        LOG_DEBUG(Service_PCTL, "called");
 
         IPC::ResponseBuilder rb{ctx, 3};
+        if (False(capability & Capability::SteroVision)) {
+            LOG_ERROR(Service_PCTL, "Application does not have SteroVision capability!");
```
Comment (lioncash) [inline src/core/hle/service/pctl/module.cpp:315]:
```suggestion
            LOG_ERROR(Service_PCTL, "Application does not have StereoVision capability!");
```

**Review by bunnei — APPROVED** [review]

**Pull request merged**